cmd/kflex/adopt/adopt.go

```diff
@@ -0,0 +1,224 @@
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	tenancyv1alpha1 "github.com/kubestellar/kubeflex/api/v1alpha1"
+	"github.com/kubestellar/kubeflex/cmd/kflex/common"
+	cont "github.com/kubestellar/kubeflex/cmd/kflex/ctx"
+	kfclient "github.com/kubestellar/kubeflex/pkg/client"
+	"github.com/kubestellar/kubeflex/pkg/util"
+)
+
+type CPAdopt struct {
+	common.CP
+	AdoptedKubeconfig             string
+	AdoptedContext                string
+	AdoptedURLOverride            string
+	AdoptedTokenExpirationSeconds int
+	SkipURLOverride               bool
+}
+
+// Adopt a control plane from another cluster
+func (c *CPAdopt) Adopt(hook string, hookVars []string, chattyStatus bool) {
+	done := make(chan bool)
+	var wg sync.WaitGroup
+	cx := cont.CPCtx{}
+	cx.Context(chattyStatus, false, false, false)
+
+	controlPlaneType := tenancyv1alpha1.ControlPlaneTypeExternal
+	util.PrintStatus(fmt.Sprintf("Adopting control plane %s of type %s ...", c.Name, controlPlaneType), done, &wg, chattyStatus)
+
+	adoptedKubeconfig := getAdoptedKubeconfig(c)
+
+	clp, err := kfclient.GetClient(c.Kubeconfig)
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "Error getting kubeflex client: %v\n", err)
+		os.Exit(1)
+	}
+	cl := *clp
+
+	clientsetp, err := kfclient.GetClientSet(c.Kubeconfig)
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "Error getting clientset: %v\n", err)
+		os.Exit(1)
+	}
+
+	if err := applyAdoptedBootstrapSecret(clientsetp, c.Name, adoptedKubeconfig, c.AdoptedContext, c.AdoptedURLOverride, c.SkipURLOverride); err != nil {
+		fmt.Fprintf(os.Stderr, "error creating adopted cluster kubeconfig: %v\n", err)
+		os.Exit(1)
+	}
+
+	cp := common.GenerateControlPlane(c.Name, string(controlPlaneType), "", hook, hookVars)
+
+	if err := cl.Create(context.TODO(), cp, &client.CreateOptions{}); err != nil {
+		fmt.Fprintf(os.Stderr, "Error creating ControlPlane object: %v\n", err)
+		os.Exit(1)
+	}
+
+	done <- true
+	wg.Wait()
+}
+
+func applyAdoptedBootstrapSecret(clientset *kubernetes.Clientset, cpName, adoptedKubeconfig, contextName, adoptedURLOverride string, skipURLOverride bool) error {
+	// Load the kubeconfig from file
+	config, err := clientcmd.LoadFromFile(adoptedKubeconfig)
+	if err != nil {
+		return fmt.Errorf("failed to load kubeconfig file %s: %v", adoptedKubeconfig, err)
+	}
+
+	// Retrieve the specified context
+	context, exists := config.Contexts[contextName]
+	if !exists {
+		return fmt.Errorf("context %s not found in the kubeconfig", contextName)
+	}
+
+	// Retrieve the associated cluster
+	cluster, exists := config.Clusters[context.Cluster]
+	if !exists {
+		return fmt.Errorf("cluster %s not found for context %s", context.Cluster, contextName)
+	}
+
+	// Construct a new kubeConfig object
+	kubeConfig := api.NewConfig()
+
+	kubeConfig.Clusters[context.Cluster] = cluster
+
+	if !skipURLOverride {
+		// Determine the server endpoint
+		endpoint := adoptedURLOverride
+		if endpoint == "" {
+			endpoint = cluster.Server
+			if !isValidServerURL(endpoint) {
```

Comment:
Why is this validity not checked for the case where `adoptedURLOverride` is not the empty string?